### ISSUE TITLE
Panic on startup if checkpoint directory is empty

### ISSUE THREAD

**Issue by brian-brazil** [prometheus/prometheus#6976]
If the checkpoint directory is present but has no files you get the following panic:

```
level=info ts=2020-03-13T08:53:02.071Z caller=head.go:584 component=tsdb msg="replaying WAL, this may take awhile"
panic: runtime error: index out of range [0] with length 0

goroutine 108 [running]:
github.com/prometheus/prometheus/tsdb/wal.NewSegmentBufReader(...)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/tsdb/wal/wal.go:830
github.com/prometheus/prometheus/tsdb/wal.NewSegmentsRangeReader(0xc00009baf8, 0x1, 0x1, 0x1a, 0xb77, 0x0, 0x0)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/tsdb/wal/wal.go:813 +0x760
github.com/prometheus/prometheus/tsdb/wal.NewSegmentsReader(...)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/tsdb/wal/wal.go:785
github.com/prometheus/prometheus/tsdb.(*Head).Init(0xc0000ee780, 0x8000000000000000, 0x0, 0x0)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/tsdb/head.go:592 +0xbc4
github.com/prometheus/prometheus/tsdb.Open(0x24dec1b, 0x5, 0x2c64440, 0xc00060b170, 0x2c9f6a0, 0xc000144460, 0xc0000a3900, 0x2, 0xc0005fb000, 0xc)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/tsdb/db.go:567 +0x5f9
github.com/prometheus/prometheus/storage/tsdb.Open(0x24dec1b, 0x5, 0x2c64440, 0xc00060b170, 0x2c9f6a0, 0xc000144460, 0xc00014a1e0, 0x0, 0x0, 0x0)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/storage/tsdb/tsdb.go:192 +0x290
main.main.func19(0x0, 0x0)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/cmd/prometheus/main.go:654 +0x31a
github.com/oklog/run.(*Group).Run.func1(0xc0000ca840, 0xc000126640, 0xc00057c000)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/vendor/github.com/oklog/run/group.go:38 +0x27
created by github.com/oklog/run.(*Group).Run
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/vendor/github.com/oklog/run/group.go:37 +0xbb
```

We should be resilient to this, and also check what can cause this to occur.

**Comment by brian-brazil** [comment]
Looking at the code, it's not possible for Prometheus to end up in this state by itself. A checkpoint is created in a tmpdir, and a segment is always created. So even a crash wouldn't do this.

**Comment by zhulongcheng** [comment]
This is the only entry point for creating checkpoint directories:
https://github.com/prometheus/prometheus/blob/4d6f9b75e230465443391ca76a4182b130eb99e9/tsdb/wal/checkpoint.go#L261-L263

Is there any error log about `rename checkpoint directory`?

Is the directory name a valid checkpoint format `checkpoint.%06d`?

**Comment by brian-brazil** [comment]
The original reporter of this issue was not able to provide me with logs.

**Comment by zhulongcheng** [comment]
We do not force sync WAL directory when [WAL is closed](https://github.com/prometheus/prometheus/blob/4d6f9b75e230465443391ca76a4182b130eb99e9/tsdb/wal/wal.go#L725).


If system crashes before flushing directory entries to disk, the directory may lose files.

How do you think about this?

**Comment by brian-brazil** [comment]
That sounds plausible, we should probably add that then before we do the rename.

**Comment by pauloconnor** [comment]
I have exactly this same issue at the moment, and this is the only reference to it I could find

```
level=info ts=2020-06-30T08:58:25.887Z caller=main.go:332 msg="Starting Prometheus" version="(version=2.14.0, branch=HEAD, revision=edeb7a44cbf745f1d8be4ea6f215e79e651bfe19)"
level=info ts=2020-06-30T08:58:25.887Z caller=main.go:333 build_context="(go=go1.13.4, user=root@df2327081015, date=20191111-14:27:12)"
level=info ts=2020-06-30T08:58:25.887Z caller=main.go:334 host_details="(Linux 4.15.0-88-generic #88-Ubuntu SMP Tue Feb 11 20:11:34 UTC 2020 x86_64 prom-net-dp-6bff67679c-hckq9 (none))"
level=info ts=2020-06-30T08:58:25.887Z caller=main.go:335 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2020-06-30T08:58:25.887Z caller=main.go:336 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2020-06-30T08:58:31.929Z caller=main.go:657 msg="Starting TSDB ..."
level=info ts=2020-06-30T08:58:31.929Z caller=web.go:496 component=web msg="Start listening for connections" address=0.0.0.0:9090
level=info ts=2020-06-30T08:58:32.252Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593403200000 maxt=1593410400000 ulid=01EBZE07WDE3F659NEPGHKVKKS
level=info ts=2020-06-30T08:58:32.256Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593410400000 maxt=1593417600000 ulid=01EBZMVZ4D451M5YTYWGPGGAEJ
level=info ts=2020-06-30T08:58:32.261Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593417600000 maxt=1593424800000 ulid=01EBZVQQYKDNBHT62EJY2T35AW
level=info ts=2020-06-30T08:58:32.265Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593424800000 maxt=1593432000000 ulid=01EC02KDMC866SKR14SE1QAEF5
level=info ts=2020-06-30T08:58:32.269Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593432000000 maxt=1593439200000 ulid=01EC09F4WCNAR75Q6FWE78N82A
level=info ts=2020-06-30T08:58:32.273Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593439200000 maxt=1593446400000 ulid=01EC0GAW4CZ92B8FS17KQNTGW3
level=info ts=2020-06-30T08:58:32.279Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593446400000 maxt=1593453600000 ulid=01EC0Q6KCDW48716V3DC7HGFB3
level=info ts=2020-06-30T08:58:32.284Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593453600000 maxt=1593460800000 ulid=01EC0Y2AMDM4F6NCRM39FQGTTZ
level=info ts=2020-06-30T08:58:32.288Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593460800000 maxt=1593468000000 ulid=01EC14Y1WCSVZSTZC5TSZTEQD1
level=info ts=2020-06-30T08:58:32.292Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593468000000 maxt=1593475200000 ulid=01EC1BSS4CSV3PRR8Z8F7VSXZW
level=info ts=2020-06-30T08:58:32.296Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593475200000 maxt=1593482400000 ulid=01EC1JNGCC8TCBXV0P77RCNDX5
level=info ts=2020-06-30T08:58:32.300Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593482400000 maxt=1593489600000 ulid=01EC1SH7MCB0KQ60MJ28KRX49C
level=info ts=2020-06-30T08:58:32.305Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593489600000 maxt=1593496800000 ulid=01EC20CYWCEE03TYWXA6522AE6
level=info ts=2020-06-30T08:58:38.638Z caller=head.go:535 component=tsdb msg="replaying WAL, this may take awhile"
panic: runtime error: index out of range [0] with length 0

goroutine 192 [running]:
github.com/prometheus/prometheus/tsdb/wal.NewSegmentBufReader(...)
	/app/tsdb/wal/wal.go:830
github.com/prometheus/prometheus/tsdb/wal.NewSegmentsRangeReader(0xc000d13af8, 0x1, 0x1, 0x21, 0x912, 0x0, 0x0)
	/app/tsdb/wal/wal.go:813 +0x760
github.com/prometheus/prometheus/tsdb/wal.NewSegmentsReader(...)
	/app/tsdb/wal/wal.go:785
github.com/prometheus/prometheus/tsdb.(*Head).Init(0xc000174000, 0x17303cf8b00, 0x0, 0x0)
	/app/tsdb/head.go:543 +0xbc4
github.com/prometheus/prometheus/tsdb.Open(0x7ffc15840407, 0xb, 0x2b4f400, 0xc00077c330, 0x2b8a220, 0xc0001cc460, 0xc0001b4b00, 0x2, 0xc00079a100, 0xc)
	/app/tsdb/db.go:562 +0x5f9
github.com/prometheus/prometheus/storage/tsdb.Open(0x7ffc15840407, 0xb, 0x2b4f400, 0xc00077c330, 0x2b8a220, 0xc0001cc460, 0xc0001d26e0, 0x0, 0x0, 0x0)
	/app/storage/tsdb/tsdb.go:194 +0x290
main.main.func19(0x0, 0x0)
	/app/cmd/prometheus/main.go:663 +0x31a
github.com/oklog/run.(*Group).Run.func1(0xc00019e420, 0xc0001b4a80, 0xc00077c210)
	/app/vendor/github.com/oklog/run/group.go:38 +0x27
created by github.com/oklog/run.(*Group).Run
	/app/vendor/github.com/oklog/run/group.go:37 +0xbb
```

```
# ls -latr prometheus-data/
total 50
-rw-r--r--  1 root root     0 Feb 18 16:35 lock
drwxr-xr-x  3 root root  4096 Jun 17 21:16 thanos
drwxr-x---  3 root root  4096 Jun 24 11:06 ..
drwxr-xr-x  3 root root    64 Jun 29 09:01 01EBZMVZ25Y41BZ3D2FQ8K69QT
drwxr-xr-x  3 root root    64 Jun 29 11:01 01EBZVQP9T7TQHJMCBFF1MHSVF
drwxr-xr-x  3 root root    64 Jun 29 13:00 01EC02KDBYXNVHN0EMDA6HWYJW
drwxr-xr-x  3 root root    64 Jun 29 15:01 01EC09F4PX5TXW3HVHDRTW0Q9V
drwxr-xr-x  3 root root    64 Jun 29 17:01 01EC0GAVVT22F3NJ6HTS0H6PNF
drwxr-xr-x  3 root root    64 Jun 29 19:01 01EC0Q6K3YXKH3HK1Y1C6MGKE1
drwxr-xr-x  3 root root    64 Jun 29 21:01 01EC0Y2AEXZKDQH0285NRD52XD
drwxr-xr-x  3 root root    64 Jun 29 23:01 01EC14Y1M2HCGQRGH0KKZK0SSF
drwxr-xr-x  3 root root  4096 Jun 30 01:01 01EC1BSRYZH056R7P0K814DYEE
drwxr-xr-x  3 root root    64 Jun 30 03:01 01EC1JNG9RJZF43RV1M9ZREG6Z
drwxr-xr-x  3 root root    64 Jun 30 05:01 01EC1SH7F2FPM0M8NYP9D7DMW4
drwxr-xr-x  3 root root  4096 Jun 30 07:01 01EC20CYPT0WAWP099D878JMXN
drwxr-xr-x  3 root root  4096 Jun 30 09:01 01EC278P28EYCH1F7X4151B6EZ
drwxr-xr-x  3 root root    90 Jun 30 09:04 wal
-rw-r--r--  1 root root 20001 Jun 30 09:06 queries.active
-rw-r--r--  1 root root    68 Jun 30 09:06 thanos.shipper.json
drwxr-xr-x 18 root root  4096 Jun 30 09:06 .
```

Let me know if I can provide any useful logs.

**Comment by codesome** [comment]
@pauloconnor do you have the logs of Prometheus shutting down?

**Comment by pauloconnor** [comment]
The only thing I can find is this:

```
level=error ts=2020-06-30T07:50:47.981Z caller=web.go:102 component=web msg="panic while serving request" client=10.202.6.192:48458 url=/metrics err="runtime error: invalid memory address or nil pointer dereference" stack="goroutine 267 [running]:
github.com/prometheus/prometheus/web.withStackTracer.func1.1(0x2b4f400, 0xc0007f1050, 0xc000192d00)
    /app/web/web.go:101 +0xb5
panic(0x208d7e0, 0x3d5a630)
    /usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/prometheus/prometheus/tsdb.(*Head).MinTime(...)
    /app/tsdb/head.go:1195
github.com/prometheus/prometheus/tsdb.newDBMetrics.func3(0x0)
    /app/tsdb/db.go:219 +0x90
github.com/prometheus/client_golang/prometheus.(*valueFunc).Write(0xc0001adec0, 0xc000655650, 0xc000d0f048, 0x3)
    /app/vendor/github.com/prometheus/client_golang/prometheus/value.go:72 +0x2b
github.com/prometheus/client_golang/prometheus.processMetric(0x2b672e0, 0xc0001adec0, 0xc000d0eed8, 0xc000d0ef08, 0x0, 0x0, 0x0)
    /app/vendor/github.com/prometheus/client_golang/prometheus/registry.go:595 +0x9b
github.com/prometheus/client_golang/prometheus.(*Registry).Gather(0xc0001d0460, 0x0, 0x0, 0x0, 0x0, 0x0)
    /app/vendor/github.com/prometheus/client_golang/prometheus/registry.go:489 +0x85e
github.com/prometheus/client_golang/prometheus/promhttp.HandlerFor.func1(0x7fb48c5ac520, 0xc0006776d0, 0xc000192f00)
    /app/vendor/github.com/prometheus/client_golang/prometheus/promhttp/http.go:126 +0x65
net/http.HandlerFunc.ServeHTTP(0xc0006013b0, 0x7fb48c5ac520, 0xc0006776d0, 0xc000192f00)
    /usr/local/go/src/net/http/server.go:2007 +0x44
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1(0x7fb48c5ac520, 0xc0006776d0, 0xc000192f00)
    /app/vendor/github.com/prometheus/client_golang/prometheus/promhttp/instrument_server.go:40 +0xbc
net/http.HandlerFunc.ServeHTTP(0xc0007952c0, 0x7fb48c5ac520, 0xc0006776d0, 0xc000192f00)
    /usr/local/go/src/net/http/server.go:2007 +0x44
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1(0x7fb48c5ac520, 0xc000677680, 0xc000192f00)
    /app/vendor/github.com/prometheus/client_golang/prometheus/promhttp/instrument_server.go:100 +0xda
net/http.HandlerFunc.ServeHTTP(0xc0007953e0, 0x7fb48c5ac520, 0xc000677680, 0xc000192f00)
    /usr/local/go/src/net/http/server.go:2007 +0x44
net/http.HandlerFunc.ServeHTTP(0xc0007b99a0, 0x7fb48c5ac520, 0xc000677680, 0xc000192f00)
    /usr/local/go/src/net/http/server.go:2007 +0x44
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerResponseSize.func1(0x7fb48c5ac520, 0xc000677630, 0xc000192f00)
    /app/vendor/github.com/prometheus/client_golang/prometheus/promhttp/instrument_server.go:196 +0xe9
net/http.HandlerFunc.ServeHTTP(0xc0007955c0, 0x7fb48c5ac520, 0xc000677630, 0xc000192f00)
    /usr/local/go/src/net/http/server.go:2007 +0x44
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2(0x7fb48c5ac520, 0xc000677630, 0xc000192f00)
    /app/vendor/github.com/prometheus/client_golang/prometheus/promhttp/instrument_server.go:76 +0xb2
net/http.HandlerFunc.ServeHTTP(0xc0007956b0, 0x7fb48c5ac520, 0xc000677630, 0xc000192f00)
    /usr/local/go/src/net/http/server.go:2007 +0x44
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1(0x2b98ae0, 0xc0006775e0, 0xc000192f00)
    /app/vendor/github.com/prometheus/client_golang/prometheus/promhttp/instrument_server.go:100 +0xda
github.com/prometheus/common/route.(*Router).handle.func1(0x2b98ae0, 0xc0006775e0, 0xc000192e00, 0x0, 0x0, 0x0)
    /app/vendor/github.com/prometheus/common/route/route.go:77 +0x296
github.com/julienschmidt/httprouter.(*Router).ServeHTTP(0xc0005e7b00, 0x2b98ae0, 0xc0006775e0, 0xc000192e00)
    /app/vendor/github.com/julienschmidt/httprouter/router.go:334 +0x95e
github.com/prometheus/common/route.(*Router).ServeHTTP(0xc0007b82e0, 0x2b98ae0, 0xc0006775e0, 0xc000192e00)
    /app/vendor/github.com/prometheus/common/route/route.go:115 +0x4c
net/http.(*ServeMux).ServeHTTP(0xc0001ac4c0, 0x2b98ae0, 0xc0006775e0, 0xc000192e00)
    /usr/local/go/src/net/http/server.go:2387 +0x1bd
github.com/opentracing-contrib/go-stdlib/nethttp.MiddlewareFunc.func5(0x2b96f20, 0xc0005f02a0, 0xc000192d00)
    /app/vendor/github.com/opentracing-contrib/go-stdlib/nethttp/server.go:140 +0x5fb
net/http.HandlerFunc.ServeHTTP(0xc0008409c0, 0x2b96f20, 0xc0005f02a0, 0xc000192d00)
    /usr/local/go/src/net/http/server.go:2007 +0x44
github.com/prometheus/prometheus/web.withStackTracer.func1(0x2b96f20, 0xc0005f02a0, 0xc000192d00)
    /app/web/web.go:106 +0xb3
net/http.HandlerFunc.ServeHTTP(0xc0008409f0, 0x2b96f20, 0xc0005f02a0, 0xc000192d00)
    /usr/local/go/src/net/http/server.go:2007 +0x44
net/http.serverHandler.ServeHTTP(0xc0005f01c0, 0x2b96f20, 0xc0005f02a0, 0xc000192d00)
    /usr/local/go/src/net/http/server.go:2802 +0xa4
net/http.(*conn).serve(0xc00003a460, 0x2ba05a0, 0xc00071c240)
    /usr/local/go/src/net/http/server.go:1890 +0x875
created by net/http.(*Server).Serve
    /usr/local/go/src/net/http/server.go:2927 +0x38e
"
```
which is then followed by `replaying WAL, this may take awhile` and the `index out of range` error

**Comment by codesome** [comment]
Huh, this error again. The ungraceful shutdown was the main issue here which might have caused the index out of range error. Related issue for the panic that you saw during shutdown: https://github.com/prometheus/prometheus/issues/7454. It is on my plate to take a look (soon).

**Comment by codesome** [comment]
The panic on the metrics during startup is fixed and will be in v2.20.0. If we could confirm if this issue was because of that, we may as well close this.
